R/input_checking_functions.R

```diff
@@ -283,19 +283,64 @@ check_time_is_sequential <- function(
     rtol = sqrt(.Machine$double.eps)
 )
 {
+    
     # only checked if differential modules are present
     if (length(differential_modules) == 0) {
         return(character())
     }
 
     no_time_variable <- !('time' %in% names(drivers))
     if (no_time_variable) {
-        return("No `time` variable found in the `drivers` dataframe.")
+        msg <- paste0(c(
+            "No `time` variable found in the `drivers` dataframe.\nSuggested fixes:", 
+            "+ Add a `time` variable spaced by `timestep`:",
+            "    drivers$time <- seq(1, nrow(drivers), timestep)",
+            "+ Add the direct module `BioCro:format_time` if drivers contain `doy` and `hour` variables."), 
+            collapse = "\n    ")
+        return(msg)
     }
 
     time <- drivers[['time']]
-    if (is.unsorted(time)) {
-        return("`time` variable is not increasing.")
+
+    if (anyNA(time)) {
+        is_na <- is.na(time)
+        count_na <- sum(is_na)
+        idx <- which(is_na)
+        msg <- paste(c(
+            sprintf("Missing values found. The `time` variable in the drivers should not contain missing values, but has %1s NA's'", count_na),
+            "To find rows with NA, call `which(is.na(time))`", 
+            "+ Check rows:",  paste0(head(idx), collapse= ', ')),    
+            collapse = '\n    ' )
+            
+        return(msg)
+    }
+
+    if (is.unsorted(time, na.rm=TRUE, strictly=TRUE)) {
+        # duplicates ? 
+        duplicates <- which(duplicated(time))
+        if (length(duplicates) > 0) {
+            msg <- paste0(c(
+               sprintf("Duplicates found. The `time` variable in the drivers contains %1s duplicate entries.", length(duplicates)),
+                "To find duplicates, call: `which(duplicated(time))`",
+                "+ Check rows:", 
+                paste0(head(duplicates), collapse= ', ')),
+               collapse ='\n    ')
+            return(msg)
+        }
+        # otherwise
+        time_sorted <- sort(time)
+        not_sorted <-time != time_sorted
+        count <- sum(not_sorted) 
+        idx <- which(not_sorted)        
+        msg <- paste0(
+            c("`time` not sorted.", 
+            sprintf("The `time` variable in the drivers should be sorted in increasing order, but %1s elements are not in order.", count), 
+            "Call `sort(drivers, by = time, ascending=TRUE)` to sort the `drivers` data.frame.", 
```

Comment:
I'm not sure this suggested fix should be in here. First, it will not work if the drivers have a different name. Second, simply re-sorting the drivers may just "sweep away" some underlying issue with them and may not be the appropriate solution.